midi_app_controller/state/state_manager.py

```diff
@@ -0,0 +1,204 @@
+        return self._connected_controller is not None
+
+    def get_available_binds(self) -> List[str]:
+        """Returns names of all binds sets suitable for current controller and app."""
+        # TODO Extract the loading logic to a separate file that will handle storage.
+        if self.selected_controller is None:
+            return []
+        all_binds = Binds.load_all_from(Config.BINDS_DIRECTORY)
+        return [
+            b.name
+            for b, _ in all_binds
+            if b.app_name == self._app_name
+            and b.controller_name == self.selected_controller.name
+        ]
+
+    def get_available_controllers(self) -> List[str]:
+        """Returns names of all controller schemas."""
+        # TODO Extract the loading logic to a separate file that will handle storage.
+        controllers = Controller.load_all_from(Config.CONTROLLERS_DIRECTORY)
+        return [c.name for c, _ in controllers]
+
+    def get_available_midi_in(self) -> List[str]:
+        """Returns names of all MIDI input ports."""
+        return self._midi_in.get_ports()
+
+    def get_available_midi_out(self) -> List[str]:
+        """Returns names of all MIDI output ports."""
+        return self._midi_out.get_ports()
+
+    def select_binds(self, name: str) -> None:
+        """Updates currently selected binds.
+
+        Does not have any immediate effect except updating the value and
+        finding path to the config file.
+        """
+        all_binds = Binds.load_all_from(Config.BINDS_DIRECTORY)
+        for binds, path in all_binds:
+            if binds.name == name:
+                self.selected_binds = SelectedItem(name, path)
+                return
+        raise Exception("Config file of selected binds not found.")
+
+    def select_controller(self, name: str) -> None:
+        """Updates currently selected controller schema.
+
+        Does not have any immediate effect except updating the value and
+        finding path to the config file.
+        """
+        all_controllers = Controller.load_all_from(Config.CONTROLLERS_DIRECTORY)
+        for controller, path in all_controllers:
+            if controller.name == name:
+                self.selected_controller = SelectedItem(name, path)
+                return
+        raise Exception("Config file of selected controller not found.")
+
+    def select_midi_in(self, name: str) -> None:
+        """Updates currently selected MIDI input port name.
+
+        Does not have any immediate effect except updating the value.
+        """
+        self.selected_midi_in = name
+
+    def select_midi_out(self, name: str) -> None:
+        """Updates currently selected MIDI output port name.
+
+        Does not have any immediate effect except updating the value.
+        """
+        self.selected_midi_out = name
+
+    def stop_handling(self) -> None:
+        """Stops handling any MIDI signals."""
+        self._midi_in.cancel_callback()
+        self._connected_controller = None
```

Comment:
Do you know if somebody else could access these ports, after we called 'stop handling', the way it is written right now?

Reply:
Yes, I think it's possible to do that.